gslib/commands/iam.py

```diff
@@ -578,16 +693,22 @@ def _PatchIam(self):
 
     # All following arguments are urls.
     for token in it:
-      patterns.append(token)
+      url_pattern_strings.append(token)
+
+    url_objects = list(map(StorageUrlFromString, url_pattern_strings))
+    _RaiseErrorIfUrlsAreMixOfBucketsAndObjects(url_objects,
```

Comment:
Shouldn't this be higher level?

Reply:
You mean higher in the file? I moved it.

Reply:
No, like, earlier in the logic before we delve into `ch`

Reply:
Ah, I see. In this case we can't really do that, since there's ch-specific logic to figure out what the URLs even are. This is related to the problem that I mentioned here: https://github.com/GoogleCloudPlatform/gsutil/pull/1727#discussion_r1261418660

Bindings that we add look like positional arguments/URLs, so we need to separate them out before we get the URLs we can pass to this method.